### PR TITLE
Bc 23520

### PR DESCRIPTION
**Pull request by razi434**

### CODE REVIEW

File: scripts/coverage.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+# run the unittests with branch coverage
+python -m pytest --cov=./heksher --cov-report=xml --cov-report=term-missing tests/unittest "$@"
+python -m pytest --cov=./heksher --cov-report=xml --cov-report=term-missing tests/blackbox/app "$@"
```
Comment (bentheiii) [inline scripts/coverage.sh:4]: append to previous coverage

File: tests/blackbox/app/test_main.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    assert heksher_service.postgres_service.is_alive()
+
+
+@mark.asyncio
+async def test_init_works(heksher_service: HeksherService):
+    resp = heksher_service.http_client.post('/api/v1/context_features', json={'context_feature': 'razi'})
+    resp.raise_for_status()
+
+    res = heksher_service.http_client.get('/api/v1/context_features')
+    res.raise_for_status()
```
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:24]: why?

File: tests/blackbox/app/test_main.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+@mark.asyncio
+async def test_get_setting_names(heksher_service: HeksherService):
+    settings = (
+        Setting("test_config", type=int, configurable_features=['a', 'b', 'c'], default_value=1,
+                metadata={"description": "test"})
+    )
+    heksher = AsyncHeksherClient(heksher_service.local_url, update_interval=60, context_features=['a', 'b', 'c'])
+    await heksher.set_as_main()  # we only want to declare settings
+    await heksher.aclose()
```
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:38]: close after the test is done

File: tests/blackbox/app/test_main.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+@mark.asyncio
+@mark.parametrize('expected', [
+    [],
+    ['a'],
+    ['b', 'a'],
+    ['a', 'b', 'c', 'd'],
+    ['e', 'f', 'g']
+])
+async def test_cf_mismatch(heksher_service, caplog, monkeypatch, expected):
+    print(f"!!!!{expected=}")
```
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:95]: Mommy mommy I found

File: tests/blackbox/conftest.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+import heksher.main_client
+from heksher.heksher_client import TemporaryClient
+from tests.blackbox.app.utils import CreateRuleParams
+
+
+@fixture(scope='session')
+def docker_client():
+    with _docker_client() as dc:
+        yield dc
```
Comment (bentheiii) [inline tests/blackbox/conftest.py:16]: not needed

File: tests/blackbox/conftest.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+
+@fixture(scope='module')
+def _heksher_service(docker_client):
+    with HeksherService.run(docker_client, heksher_startup_context_features="a;b;c", remove=True) as service:
```
Comment (bentheiii) [inline tests/blackbox/conftest.py:21]:
```suggestion
    with HeksherService.run(docker_client, heksher_startup_context_features="a;b;c") as service:
```

File: tests/blackbox/conftest.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+@fixture(scope="function")
+def add_rules(heksher_service: HeksherService):
+    async def _add_rules(rules: Sequence[CreateRuleParams]):
+        for rule in rules:
+            resp = (heksher_service.http_client.post('/api/v1/rules', content=rule.json()))
+            if resp.is_error:
+                raise Exception(resp.content)
+
+    yield _add_rules
```
Comment (bentheiii) [inline tests/blackbox/conftest.py:46]:
```suggestion
    return _add_rules
```

File: heksher/setting_type.py [hunk 8]
```diff
@@ -119,6 +130,10 @@ def convert(self, x) -> Conversion[F]:
                 ret |= member
         return Conversion(ret, coercions)
 
+    def convert_to_heksher(self, x: F) -> Any:
+        members = [(x & self.type_(i)).name for i in self.type_]
+        return [x for x in members if x]
```
Comment (razi434) [inline heksher/setting_type.py:135]: Didn't know how to avoid returning None values in the list

File: heksher/setting_type.py [hunk 9]
```diff
@@ -62,6 +63,13 @@ def convert(self, x) -> Conversion[T]:
         """
         pass
 
+    @abstractmethod
+    def convert_to_heksher(self, x: T) -> Any:
+        """
+
+        """
```
Comment (bentheiii) [inline heksher/setting_type.py:70]: add documentation

File: heksher/setting_type.py (hunk 8 above)
Comment (bentheiii) [inline heksher/setting_type.py:135]:
```suggestion
        return [flag.name for flag in self.type_ if x & flag]
```

File: heksher/setting_type.py [hunk 10]
```diff
@@ -163,18 +181,22 @@ def __init__(self, inner: SettingType[T]):
...
-                conversion = self.inner.convert(v)
+                conversion = self.inner.convert_from_heksher(v)
             except TypeError as e:
                 coercions.append(f'failed to convert element {i}: {e!r}')
             else:
                 values.append(conversion.value)
                 coercions.extend(f'element {i}: {c}' for c in conversion.coercions)
-        return Conversion(tuple(values), coercions)
+        ret = Conversion(tuple(values), coercions)
+        return ret
```
Comment (bentheiii) [inline heksher/setting_type.py:196]: ?

File: heksher/setting_type.py [hunk 11]
```diff
@@ -163,18 +181,22 @@ def __init__(self, inner: SettingType[T]):
...
+
+    def convert_to_heksher(self, x: T) -> Any:
+        return [self.inner.convert_to_heksher(v) for i, v in enumerate(x)]
```
Comment (bentheiii) [inline heksher/setting_type.py:199]:
```suggestion
        return [self.inner.convert_to_heksher(v) for v in x]
```

File: scripts/coverage.sh (hunk 1 above)
Comment (bentheiii) [inline scripts/coverage.sh:4]: add `-a` flag

File: tests/blackbox/app/test_main.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+@mark.asyncio
+async def test_health_check(heksher_service):
+    heksher_service.http_client.get('/api/health').raise_for_status()
+    assert heksher_service.postgres_service.is_alive()
+
+
+@mark.asyncio
+async def test_init_works(heksher_service: HeksherService):
+    resp = heksher_service.http_client.post('/api/v1/context_features', json={'context_feature': 'razi'})
+    resp.raise_for_status()
```
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:21]: U2U
```suggestion
    heksher_service.http_client.post('/api/v1/context_features', json={'context_feature': 'razi'}).raise_for_status()
```

File: tests/blackbox/app/test_main.py (hunk 2 above)
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:24]: same

File: tests/blackbox/app/test_main.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+@mark.asyncio
+async def test_regular_update(heksher_service, monkeypatch, add_rules):
+    setting = Setting('cache_size', type=int, configurable_features=['b', 'c'], default_value=50)
+
+    client = AsyncHeksherClient(heksher_service.local_url, 0.02, ['a', 'b', 'c'])
+    await client.set_as_main()
+    client.track_contexts(b='0')
+
+    await add_rules([CreateRuleParams(setting='cache_size', feature_values={'b': '0'}, value=100)])
+    await sleep(1)
```
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:81]: reload instead

File: tests/blackbox/app/test_main.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    assert setting.get(b='5', c='') == 100
+    await client.aclose()
+
+
+@mark.asyncio
+async def test_regular_update(heksher_service, monkeypatch, add_rules):
+    setting = Setting('cache_size', type=int, configurable_features=['b', 'c'], default_value=50)
+
+    client = AsyncHeksherClient(heksher_service.local_url, 0.02, ['a', 'b', 'c'])
+    await client.set_as_main()
```
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:77]: you should do `async with` in these cases

File: tests/blackbox/app/test_main.py (hunk 4 above)
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:95]: why is this still here?

File: tests/blackbox/app/utils.py [hunk 15]
```diff
@@ -0,0 +1,10 @@
+from typing import Any, Dict
+
+from pydantic import BaseModel, Field
+
+
+class CreateRuleParams(BaseModel):
```
Comment (bentheiii) [inline tests/blackbox/app/utils.py:6]: prefer dataclass for non-model work

File: tests/blackbox/conftest.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+@fixture(scope='module')
+def _heksher_service(docker_client):
+    with HeksherService.run(docker_client, heksher_startup_context_features="a;b;c") as service:
+        yield service
+
+
+@fixture(scope='function')
+async def heksher_service(_heksher_service: HeksherService):
+    yield _heksher_service
+    # input("!!!")
```
Comment (bentheiii) [inline tests/blackbox/conftest.py:21]: delete

File: tests/blackbox/conftest.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+@fixture(scope='session')
+def event_loop(request):
+    loop = asyncio.get_event_loop_policy().new_event_loop()
+    yield loop
+    loop.close()
+
+
+@fixture(scope="function")
+def add_rules(heksher_service: HeksherService):
+    async def _add_rules(rules: Sequence[CreateRuleParams]):
```
Comment (bentheiii) [inline tests/blackbox/conftest.py:34]:
```suggestion
    async def _add_rules(rules: Iterable[CreateRuleParams]):
```

Comment (bentheiii) [inline tests/blackbox/conftest.py:34]: nothing about this is async

File: tests/unittest/test_setting_type.py [hunk 18]
```diff
@@ -17,10 +17,13 @@ class A(Enum):
...
-    assert t.convert(0) == Conversion(A.a)
-    assert t.convert("c") == Conversion(A.c)
+    assert t.convert_from_heksher(0) == Conversion(A.a)
+    assert t.convert_from_heksher("c") == Conversion(A.c)
     with raises(TypeError):
-        assert t.convert("b")
+        assert t.convert_from_heksher("b")
+
+    assert t.convert_to_heksher(A.a) == 0
+    assert t.convert_to_heksher(A.c) == 'c'
```
Comment (bentheiii) [inline tests/unittest/test_setting_type.py:26]: put these in a different test

File: tests/unittest/test_setting_type.py [hunk 19]
```diff
@@ -32,9 +35,13 @@ class B(IntFlag):
...
-    assert t.convert(["a", "c"]) == Conversion(B.a, ['server sent a flag value not found in the python type (c)'])
-    assert t.convert([]) == Conversion(B(0), [])
-    assert t.convert(["a", "b"]) == Conversion(B.a | B.b, [])
+    assert t.convert_from_heksher(["a", "c"]) == Conversion(B.a, [
+        'server sent a flag value not found in the python type (c)'])
+    assert t.convert_from_heksher([]) == Conversion(B(0), [])
+    assert t.convert_from_heksher(["a", "b"]) == Conversion(B.a | B.b, [])
+
+    assert t.convert_to_heksher(B.b) == ["b"]
+    assert t.convert_to_heksher(B.a | B.b) == ["a", "b"]
```
Comment (bentheiii) [inline tests/unittest/test_setting_type.py:44]: same, also for other types

File: heksher/setting_type.py (hunk 10 above)
Comment (razi434) [reply heksher/setting_type.py:196]: Oy vey, I did it for printing and forgot to change back

File: heksher/clients/async_client.py [hunk 20]
```diff
@@ -247,8 +250,7 @@ async def get_settings(self) -> Dict:
         """
         List all the settings in the service
         """
-        response = await self._http_client.get('/api/v1/settings', params=orjson.dumps(
-            {'include_additional_data': True}))
+        response = await self._http_client.get('/api/v1/settings', params={'include_additional_data': 'True'})
```
Comment (bentheiii) [inline heksher/clients/async_client.py:253]: add to changelog

File: heksher/setting.py [hunk 21]
```diff
@@ -162,7 +162,7 @@ def to_v1_declaration_body(self) -> Dict[str, Any]:
             'type': self.type.heksher_string(),
             'metadata': self.metadata,
             'alias': self.alias,
-            'default_value': self.default_value,
+            'default_value': self.type.convert_to_heksher(x=self.default_value),
```
Comment (bentheiii) [inline heksher/setting.py:165]:
```suggestion
            'default_value': self.type.convert_to_heksher(self.default_value),
```

File: tests/blackbox/app/test_main.py [hunk 22]
```diff
@@ -0,0 +1,427 @@
+from asyncio import sleep
```
Comment (bentheiii) [inline tests/blackbox/app/test_main.py:1]: try to split this file...

File: tests/blackbox/app/utils.py [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+from dataclasses import dataclass, field
+from typing import Any, Dict, Optional
+
+
+@dataclass
+class CreateRuleParams:
+    setting: str  # The name of the configuration
+    feature_values: Dict[str, str]
+    value: Any
+    metadata: Optional[Dict[str, str]] = field(default_factory=dict)
```
Comment (bentheiii) [inline tests/blackbox/app/utils.py:11]:
```suggestion
    metadata: Dict[str, Any] = field(default_factory=dict)
```

File: heksher/setting_type.py [hunk 24]
```diff
@@ -62,6 +62,17 @@ def convert(self, x) -> Conversion[T]:
         """
         pass
 
+    @abstractmethod
+    def convert_to_heksher(self, x: T) -> Any:
```
Comment (bentheiii) [inline heksher/setting_type.py:66]: changelog

File: CHANGELOG.md [hunk 25]
```diff
@@ -1,4 +1,8 @@
 # heksher-py Changelog
+## Next
+### Changed
+* in get_settings(), changed the params from orjson.dumps to an actual dictionary
+* convert() in setting_type.py split into convert_from heksher() and convert_to_heksher()
```
Comment (bentheiii) [inline CHANGELOG.md:5]: Changelogs are for external users, not for internal implementation

```suggestion
### Fixed
* in `get_settings`: getting settings with metadata works
* creating a flags setting now works
```

File: tests/unittest/util.py [hunk 26]
```diff
@@ -1,12 +1,15 @@
...
 
 @contextmanager
-def assert_logs(caplog, level):
+def assert_logs(caplog, level, pattern: Union[Pattern[str], str]):
+    pattern = re.compile(pattern)
     caplog.clear()
     with caplog.at_level(level):
         yield
-        assert caplog.records
+        assert any([pattern.fullmatch(record.msg) for record in caplog.records])
```
Comment (bentheiii) [inline tests/unittest/util.py:12]:
```suggestion
        assert any(pattern.fullmatch(record.msg) for record in caplog.records)
```

**Review by bentheiii — APPROVED** [review]

**Pull request merged**